File: Q2/src/main/java/com/yourorganization/Transport/Ship.java
package com.yourorganization.Transport; // Package for transport modes

import java.util.List;



import java.util.ArrayList; // Imports for list handling

public class Ship extends Transport {
    private List<String> seaRoutes; // List of sea routs the ship travels

    // Constructor for initializing Ship
    public Ship(String id,double speed, double costPerKm, List<String> seaRoutes) {
        super(id, speed, costPerKm); // Invoking Transport's constructor
        this.seaRoutes = new ArrayList<>(seaRoutes); // Assigning sea routes
    }

    // Overriding getRoutePlan from the Transport class
    @Override
    public List<String> getRoutePlan() {
        return seaRoutes; 
    }

    // Getter and setter for seaRoutes
    public List<String> getSeaRoutes() {
        return seaRoutes;
    }

    public void setSeaRoutes(List<String> seaRoutes) {
        this.seaRoutes = seaRoutes;
    }
}

File: Q3_Cpp/src/Transport/Ship.cpp
#include "../../include/Transport/Ship.h"
#include <sstream>
#include <algorithm>
#include <iterator>

// Constructor for initializing Ship
Ship::Ship(std::string id, double speed, double costPerKm, const std::vector<std::string>& seaRoutes)
    : Transport(id, speed, costPerKm), seaRoutes(seaRoutes) {
}

// Overriding getRoutePlan from the Transport class
std::string Ship::getRoutePlan() {
    std::ostringstream oss;
    if (!seaRoutes.empty()) {
        std::copy(seaRoutes.begin(), seaRoutes.end() - 1, std::ostream_iterator<std::string>(oss, ", "));
        oss << seaRoutes.back();
    }
    return "Ship Route: " + oss.str(); // Concatenating sea routes
}

// Getter and setter for seaRoutes
std::vector<std::string> Ship::getSeaRoutes() const {
    return seaRoutes;
}

void Ship::setSeaRoutes(const std::vector<std::string>& seaRoutes) {
    this->seaRoutes = seaRoutes;
}
